checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -307,14 +308,29 @@ private void refineEq(
     }
 
     /**
-     * If lengthAccess node is an sequence length field or method access and the other node is less
-     * than or equal to that sequence length, then refine the other nodes type to less than the
-     * sequence length.
+     * If lengthAccess node is an sequence length field or method access (optionally with a literal
+     * offset subtracted) and the other node is less than or equal to that sequence length (minus
+     * the offset), then refine the other nodes type to less than the sequence length (minus the
+     * offset).
      */
     private void refineNeqSequenceLength(
             Node lengthAccess, Node otherNode, AnnotationMirror otherNodeAnno, CFStore store) {
 
         Receiver receiver = null;
+        // If lengthAccess is "receiver.length - c" where c is an integer literal, stores c
+        // into lengthOffset
+        int lengthOffset = 0;
+        if (lengthAccess instanceof NumericalSubtractionNode) {
+            NumericalSubtractionNode subtraction = (NumericalSubtractionNode) lengthAccess;
+            Node offsetNode = subtraction.getRightOperand();
+            if (offsetNode instanceof IntegerLiteralNode) {
```

Comment:
Instead of only checking for an `IntegerLiteralNode`, why not use the Value Checker and check if there's a single estimate for the value? That will be even more general.